### PR TITLE
Update code review documentation with additional information

### PR DESCRIPTION
**Pull request by ccostino**
This change set adds several additions to our code review documentation that expand on the following concepts:

- Benefits of code reviews
- What to look for when reviewing code

There are several document formatting updates and grammatical fixes made as well.

### CODE REVIEW

File: pages/code-review.md [hunk 1]
```diff
@@ -14,28 +14,61 @@ channel!
...
+organization's workflow, there are several obstacles you might run into:
+
+- Code reviews require intense concentration when they are to be done right,
+  which in turn requires additional time that will have an impact on story
+  effort estimates in a sprint.
+- Introducing the idea of writing tests and requiring them in order for a code
+  review to pass will take time and commitment from all team members if they
+  have not already bought into the benefits of TDD.
+- Additional training may be needed for folks who are unfamiliar with
+  what linters and/or testing harnesses.
```
Comment (pkarman) [inline pages/code-review.md]: this is an incomplete sentence

File: pages/code-review.md [hunk 2]
```diff
@@ -14,28 +14,61 @@ channel!
...
+goals when reviewing code are to:
 
 - expose bugs before they make it to production;
 - ensure a consistent quality in the codebase;
 - create an environment for sharing knowledge and developing skills;
+- cross-pollination of debugging skills when problems arise;
+- cultivation of the ability to critique one’s own code more strongly;
 - encourage open communication between the entire team.
 
+## Potential obstacles to be mindful of
```
Comment (jessieay) [inline pages/code-review.md]: would it be possible to frame this in terms of solutions rather than problems? (aka - include these in "tips for a successful review" or another category that makes sense)

File: pages/code-review.md [hunk 3]
```diff
@@ -14,28 +14,61 @@ channel!
...
+  have not already bought into the benefits of TDD.
+- Additional training may be needed for folks who are unfamiliar with
+  what linters and/or testing harnesses.
+- Writing a style guide that is firm but reasonable requires time and a strong
+  commitment from all team members to adhere to and hold each other accountable
+  for.
+- Maintaining a well-organized code base requires strict discipline from all
+  team members.
+
+In order to overcome these obstacles be sure to maintain a positive but
```
Comment (jessieay) [inline pages/code-review.md]: Since we have bullets above, maybe we should stay with that pattern and make everything a bullet point rather than a combo of bullets and paragraphs?

File: pages/code-review.md [hunk 4]
```diff
@@ -14,28 +14,61 @@ channel!
...
+For a code review to be successful, we all need to be on the same page. To
+help accomplish this, everyone should follow these tips.
 
 ### For everyone
 
 - Remember that we're reviewing the code, not the author. Nothing should be
   given or taken taken personally.
-- There is often more than one way to approach a solution. Discuss tradeoffs and
-  reach a resolution quickly.
+- Understand the difference between "this isn't how I would do it" vs.
```
Comment (jessieay) [inline pages/code-review.md]: how would understand this? Can this recommendation be more specific?

File: pages/code-review.md [hunk 5]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
+    - [ ] Are there any glaring syntax errors that were missed?
+    - [ ] Are language constructs being utilized properly?
+    - [ ] Are there no unecessary premature optimizations?
+    - [ ] Are any frameworks/libraries being used leveraged properly?
+    - [ ] Are the changes clean and would they pass through a linter?
+    - [ ] Are there useful comments/documentation where needed?
+    - [ ] Is the commit message well-formed with succinct details about
+          what the change is intended for?
+
+In adition to this checklist, you should also pull down the actual changes
```
Comment (jessieay) [inline pages/code-review.md]: spelling: addition

File: pages/code-review.md [hunk 6]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
+and run them manually to put them through their paces.  Does the code
```
Comment (jessieay) [inline pages/code-review.md]: I am not sure that I agree with this recommendation. I am doing this now because it is a practice on my team, but historically I trust that the developer working on the problem has done this and that the automated tests, if they are good, will catch any problems with functionality

File: pages/code-review.md (hunk 1 above)
Comment (ccostino) [inline pages/code-review.md]: Yup, extraneous word left in there, thanks!

File: pages/code-review.md (hunk 2 above)
Comment (ccostino) [inline pages/code-review.md]: Yes, good idea!

File: pages/code-review.md (hunk 4 above)
Comment (ccostino) [inline pages/code-review.md]: Ah, good point.  I will either clarify with an example or consolidate with the related bullet.

File: pages/code-review.md (hunk 3 above)
Comment (ccostino) [inline pages/code-review.md]: That's true - I wanted a bit of discussion around these points but you're right, the rest of the document is largely a collection of bullet points and checklists.  I'll edit some more and see where I end up.

File: pages/code-review.md (hunk 5 above)
Comment (ccostino) [inline pages/code-review.md]: D'oh!

File: pages/code-review.md (hunk 6 above)
Comment (ccostino) [inline pages/code-review.md]: I disagree that automated tests alone will catch any problems with functionality.  What if the code provides many good tests that all pass, yet still ultimately ends up being incorrect?  What if the functionality itself misses the mark despite being well-architected and working correctly?

The Dolores Landingham bot is a great recent example of this, where the functionality of selecting a time for messages to be sent is straightforward in theory but difficult to implement correctly.  Recall how many iterations it took of running the code through its paces to get it right regarding the time selection and time zone checking, and it's still not quite correct despite numerous tests accounting for a variety of edge cases! :-)  The day check we had to add on account of using UTC in particular, that is not something I would have accounted for at all had the code not been run outside of initial development - it just didn't even enter my purview while working on it initially.

I trust that the developer has run their code as well, but there will be some level of confirmation bias to contend with.  While the tests might cover a wide variety of use/edge cases they are no guarantee that the functionality itself is correct.  What if the tests are testing the wrong thing(s)?  What if the functionality developed is technically sound and correct implementation-wise, but misses the mark in delivering the correct end result or introduces some unintended side effect that the developer isn't aware of (perhaps some other sub-system is touched without their knowledge)?

Running the code with a fresh set of eyes is immensely helpful in uncovering these these harder-to-find issues and verifying that the code does work as expected.  I also think this exercise is helpful to the code reviewer in understanding what the code is doing and provides a broader context around how the code and and project are evolving, thus contributing to overall depth of knowledge.

I'm definitely curious why you're not sure about this recommendation, what are your thoughts?

File: pages/code-review.md (hunk 4 above)
Comment (ccostino) [inline pages/code-review.md]: Yeah, this is already covered by the two bullet points surrounding it with better language, I'm taking it out.

File: pages/code-review.md (hunk 6 above)
Comment (monfresh) [inline pages/code-review.md]: I agree with Carlo that a second pair of eyes is very helpful. It's very human to miss things during development and testing. Having said that, not all pull requests will warrant manual QA, so I think the reviewer should use their best judgement.

File: pages/code-review.md [hunk 7]
```diff
@@ -14,28 +14,61 @@ channel!
 
 ## Why reviews?
 
-Code reviews are incredibly important step to the development process. Our goals
-when reviewing code are to:
+Code reviews are incredibly important step to the development process. Our
```
Comment (monfresh) [inline pages/code-review.md]: How about "Code reviews are an incredibly important part of the development process"?

File: pages/code-review.md [hunk 8]
```diff
@@ -14,28 +14,61 @@ channel!
...
+goals when reviewing code are to:
 
 - expose bugs before they make it to production;
 - ensure a consistent quality in the codebase;
 - create an environment for sharing knowledge and developing skills;
+- cross-pollination of debugging skills when problems arise;
+- cultivation of the ability to critique one’s own code more strongly;
```
Comment (monfresh) [inline pages/code-review.md]: To make the bullet points more consistent grammatically, they should start with verbs, like the first three points. So, "cross-pollinate debugging skills..." and "cultivate the ability...". Also, I think "one's own" is redundant. "one's code" reads better to me.

File: pages/code-review.md [hunk 9]
```diff
@@ -14,28 +14,61 @@ channel!
 
 ## Why reviews?
 
-Code reviews are incredibly important step to the development process. Our goals
-when reviewing code are to:
+Code reviews are incredibly important step to the development process. Our
+goals when reviewing code are to:
 
 - expose bugs before they make it to production;
 - ensure a consistent quality in the codebase;
```
Comment (monfresh) [inline pages/code-review.md]: How about "ensure consistent code quality"?

File: pages/code-review.md [hunk 10]
```diff
@@ -14,28 +14,61 @@ channel!
...
+
+In order to overcome these obstacles be sure to maintain a positive but
+persistent attitude around the benefits of code reviews.  Continue to work with
+your colleagues and managers to demonstrate the benefits of code reviews by
+exhibiting patience with them and finding ways that showcase how code reviews
+can have a direct impact on improving their work.
+
+By maintaining a positive attititude, leading by example, and demonstrating an
+honest willingess to work with others you will begin to see people come
+around to the idea that code reviews are beneficial for everyone involved.
```
Comment (cmc333333) [inline pages/code-review.md]: Do these last two paragraphs belong in the "successful review" section?

File: pages/code-review.md [hunk 11]
```diff
@@ -90,22 +124,23 @@ Before seeking a review, you should be able to check off each of the following:
...
-    - [ ] Project documentation has been updated (including the "Unreleased" section of
-          the CHANGELOG)
+    - [ ] Project documentation has been updated (including the "Unreleased"
+          section of the CHANGELOG)
 
 ### Opening a PR
 
 Once your code meets the checks above, open a Pull Request for your new
-feature or bugfix.Remember that the more information you provide to reviews,
+feature or bugfix. Remember that the more information you provide to reviews,
```
Comment (cmc333333) [inline pages/code-review.md]: Should this be "you provide to reviewers"?

File: pages/code-review.md [hunk 12]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
     - [ ] Is the code as efficient as possible?
-    - [ ] Is the usage of each function clear?
+    - [ ] Is the usage of each function/class clear?
     - [ ] Have edge cases been considered and tested for?
-
-Use this bookmarklet code to add this checklist to your Pull Request comments: [Code Review Checklist](https://gist.github.com/cfarm/a4174fe6f775353a3115)
+    - [ ] Does the code represent a logical unit of work?
+    - [ ] Are there any glaring syntax errors that were missed?
+    - [ ] Are language constructs being utilized properly?
+    - [ ] Are there no unecessary premature optimizations?
```
Comment (cmc333333) [inline pages/code-review.md]: This would appear (on it's face) to be in conflict with "Is the code as efficient as possible"

File: pages/code-review.md [hunk 13]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
+    - [ ] Are any frameworks/libraries being used leveraged properly?
+    - [ ] Are the changes clean and would they pass through a linter?
```
Comment (cmc333333) [inline pages/code-review.md]: "clean" is a very vague word. Is there a way to expand on it or combine it with the "glaring syntax errors" and/or "Are functions/classes as simple as possible"?

File: pages/code-review.md [hunk 14]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
+    - [ ] Are there useful comments/documentation where needed?
+    - [ ] Is the commit message well-formed with succinct details about
+          what the change is intended for?
```
Comment (cmc333333) [inline pages/code-review.md]: I disagree with this requirement. While it'd be nice to have good commit messages, I don't think it's worth stressing and don't want to promote rewriting history.

File: pages/code-review.md (hunk 6 above)
Comment (cmc333333) [inline pages/code-review.md]: I agree with @jessieay - not all projects _want_ to use this flow. Some projects might prefer to QA at the end of a sprint, or perhaps before code gets merged into some sort of `production` branch. We shouldn't dictate that here.

File: pages/code-review.md [hunk 15]
```diff
@@ -14,28 +14,61 @@ channel!
...
+  which in turn requires additional time that will have an impact on story
+  effort estimates in a sprint.
+- Introducing the idea of writing tests and requiring them in order for a code
+  review to pass will take time and commitment from all team members if they
+  have not already bought into the benefits of TDD.
+- Additional training may be needed for folks who are unfamiliar with
+  what linters and/or testing harnesses.
+- Writing a style guide that is firm but reasonable requires time and a strong
+  commitment from all team members to adhere to and hold each other accountable
+  for.
```
Comment (monfresh) [inline pages/code-review.md]: I personally don't see this as an obstacle, and don't think we would need to write a style guide from scratch. Ruby and Python (and probably other languages) have established community-based style guides that we can adopt and tweak to our needs if necessary. There is also an effort underway to recommend the use of such style guides and automated checks via services like Hound, which reduces the burden on reviewers and helps team members adhere to the style guide.

File: pages/code-review.md [hunk 16]
```diff
@@ -114,12 +149,26 @@ When reviewing code, you should be able to check off each of the following:
...
+    - [ ] Are there useful comments/documentation where needed?
+    - [ ] Is the commit message well-formed with succinct details about
+          what the change is intended for?
+
+In adition to this checklist, you should also pull down the actual changes
+and run them manually to put them through their paces.  Does the code
+actually do what you expect it to do after reading through it?
+
+Use this bookmarklet code to add this checklist to your Pull Request comments:
+[Code Review Checklist](https://gist.github.com/cfarm/a4174fe6f775353a3115)
```
Comment (cmc333333) [inline pages/code-review.md]: This is redundant with the earlier link, hey? Also, do you think we should link there at all, if we now have a different checklist?

**Comment by cmc333333** [comment]
Thanks for allowing us to collectively hash much of this out in your PR, @ccostino!

File: pages/code-review.md (hunk 14 above)
Comment (monfresh) [inline pages/code-review.md]: I disagree with @cmc333333. I think detailed commit messages are extremely important. I strongly believe that every commit message must explain what the change is and why the change is necessary. Some changes will require more detail than others. A small typo fix, for example, can probably be described in just the commit title. Other more substantial changes need justification so that anyone who comes across that commit at any point in time can easily understand what happened and why.

I also believe in adhering to style rules for commit messages, such as subjects no longer than 50 characters, using the imperative mood in the subject, no period at the end of the subject.

Here are some relevant links:
http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
http://chris.beams.io/posts/git-commit/
https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message

Comment (cmc333333) [inline pages/code-review.md]: @monfresh those are reasons to _write_ a good commit message, not why the messages should be changed after the fact. If you wrote a bad message, you wrote a bad message; now it's part of the history of the project and you will look like a fool for all of history.

I don't think it's a worthwhile venture to ask a dev to change their commit messages -- we need to stop the meta analysis at some point to remain productive.

Comment (monfresh) [inline pages/code-review.md]: I still disagree. If the purpose of a code review is to make sure bad code is not merged, why not also make sure bad commit messages don't make it either? We are talking about commit messages made during a PR, before merging into the main branch. I don't have any objections to using `git commit --amend` to rewrite the commit message before it gets merged.

Comment (cmc333333) [inline pages/code-review.md]: >  If the purpose of a code review is to make sure bad code is not merged, why not also make sure bad commit messages don't make it either?

We agree that, by and large, commit messages aren't as important as the code itself. This is one of the core principles of agile, after all -- "Working software over comprehensive documentation". So the real debate is over how do we treat this less important artifact.

In my opinion, pull requests are where we review code, not team norms. If you think another dev isn't including enough detail in their changesets, that's a retro discussion, not a pull request discussion.

Comment (jessieay) [inline pages/code-review.md]: As a code reviewer, it is very helpful to have a detailed commit message / PR description in order to understand the problem that the code you are reviewing is supposed to solve (aka - does the solution match the problem?). For this reason, I would say that a detailed commit message is important. 

That being said, maybe we should take this bullet out and discuss in a separate thread. Seems like there is a lot to discuss on this point and it would be nice to have the convo in once place rather than within this much larger PR

**Comment by ccostino** [comment]
@cmc333333:  Absolutely, there's a lot of great discussion here and I am looking forward to diving into these responses further when I have a moment again!

File: pages/code-review.md (hunk 7 above)
Comment (ccostino) [inline pages/code-review.md]: Yes, that reads better!

File: pages/code-review.md (hunk 8 above)
Comment (ccostino) [inline pages/code-review.md]: Those also read better, I will adjust!

File: pages/code-review.md (hunk 9 above)
Comment (ccostino) [inline pages/code-review.md]: +1, that is succinct.

File: pages/code-review.md (hunk 10 above)
Comment (ccostino) [inline pages/code-review.md]: I am taking these out based on @jessieay's feedback earlier, I consolidated a couple of the points as additional bullets in the "Tips for a successful review" section. :-)

File: pages/code-review.md (hunk 11 above)
Comment (ccostino) [inline pages/code-review.md]: In looking at that sentence more closely (I just added the space after the period), yes, looks like it was incorrect.

File: pages/code-review.md (hunk 12 above)
Comment (ccostino) [inline pages/code-review.md]: Yeah, these seem like two-sides of the same coin to me now that you mention it.  On the one-hand code that is going to be merged should be efficient, but it also shouldn't go out of its way to address things that are unnecessary.  There are times when developers (myself included!) try and account for things in advance of them happening; defensive programming, so-to-speak.

On the surface these kinds of changes or improvements (in the eyes of the developer) look innocuous but in reality they could actually make things more difficult to refactor or extend later on, and I wanted to call this out explicitly.  I suppose this particular point could also fall under addressing scope creep, too.  I see how the bullet you mentioned is in conflict with this though, so I will attempt to re-word or just remove it.

File: pages/code-review.md (hunk 13 above)
Comment (ccostino) [inline pages/code-review.md]: Ah, good point.  "Clean" in this case is referring to code style.

File: pages/code-review.md (hunk 6 above)
Comment (ccostino) [inline pages/code-review.md]: I understand the concern that not all pull requests will warrant needing to go through the step of running things manually.  Truly dead simple changes should not soak up a lot of time, I agree!

I also understand the concern around prescribing what all teams **must** do as a part of their code reviews.  However, I am looking at the development guide through the lens of helping establish good baseline practices at 18F and leading by example.  In my eyes that translates into performing the necessary due diligence to ensure high quality work is consistently delivered and clearly documenting what that process entails.

In the case of code reviews, I feel that pulling down the changes; making sure that they do merge seamlessly (especially if there are any schema or data migrations involved); and actually using the code, even if just a high level pass, is an important part of the review process and helps immensely in ensuring high quality work.  Speaking from personal experience, I have been bitten far more times than I care to admit by the "did you actually run the code?" question when it comes to signing off or giving the thumbs up on someone else's work only to have something fail in production.

All of that said, if the sticking point is about what any given team *should* do versus what they *ought* to do to as a part of the code review process, what if the language was changed to say "At you or your team's discretion you may also consider pulling down the changes yourself and running through the code locally as a part of the review," or something to that effect?  Would that account for situations where this step is indeed overkill and preserve the autonomy of the teams, but still call out that this is indeed a useful exercise to consider?

File: pages/code-review.md (hunk 14 above)
Comment (ccostino) [inline pages/code-review.md]: @jessieay:  I think you're right and that this warrants a separate discussion that would be useful to have.  For now I'll take the bullet out, but I see the commit subject and messages as a part of any documentation that should be shipped with the code.  They are the single best fighting chance someone has in identifying and understanding why a change was made, especially when hunting down a defect or regression.

File: pages/code-review.md (hunk 15 above)
Comment (ccostino) [inline pages/code-review.md]: Duly noted!  As mentioned above, this is all being removed. :-)

File: pages/code-review.md (hunk 16 above)
Comment (ccostino) [inline pages/code-review.md]: Agreed, I don't see a need for it since we have our own checklist here now.

File: pages/code-review.md (hunk 6 above)
Comment (jessieay) [inline pages/code-review.md]: I would put this in the same category as the commit message point - let's remove and have a separate discussion on QA!

Comment (ccostino) [inline pages/code-review.md]: Sounds like a plan! :-)

**Comment by ccostino** [comment]
Lots of great discussion and feedback here, I am going to incorporate it all in another commit later tonight for further review, thank you everyone!

File: pages/code-review.md (hunk 12 above)
Comment (ccostino) [inline pages/code-review.md]: Opting to stick with just the original line about being efficient as possible after further thought.

File: pages/code-review.md (hunk 13 above)
Comment (ccostino) [inline pages/code-review.md]: Which is covered by the line "Do the changes respect the project's existing style?" - removing.

**Comment by jessieay** [comment]
This looks awesome! 

Hopefully from here on out we can keep PRs to this repo smaller so people can debate issues one at a time. 

Thank you, Carlo, for dealing with a zillion different threads on this PR, I think that after it is merged we will have a good base to work off of !! :ship:

**Comment by ccostino** [comment]
Thanks, @jessieay!  Glad to have fielded a bunch of things and work everything in that made sense, and there's plenty of material to move forward with on future improvements/guidelines. :-)

**Comment by cmc333333** [comment]
Consensus met, merge is go!
